lib/Runtime/Base/ObjectTypeCache.cpp

```diff
@@ -0,0 +1,51 @@
+//-------------------------------------------------------------------------------------------------------
+// Copyright (C) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
+
+#include "RuntimeBasePch.h"
+#include "Types/PathTypeHandler.h"
+
+namespace Js
+{
+
+DynamicType* ObjectTypeCache::FindOrCreateType(_In_ RecyclableObject* prototype)
+{
+    // For fast lookup, cache index is a function of prototype address
+    const uint index = (uint)(((uintptr_t)prototype) >> PolymorphicInlineCacheShift) & (ObjectTypeCache::MaxCachedTypes - 1);
+
+    RecyclerWeakReference<RecyclableObject>* cachedWeakProto = cache[index].prototype;
+
+    // Check for nullptr, because the cache will not have a weak ref yet if this is first access
+    if (cachedWeakProto != nullptr)
+    {
+        Assert(cache[index].type != nullptr);
+
+        DynamicType* cachedType = cache[index].type->Get();
+        RecyclableObject* cachedProto = cachedWeakProto->Get();
+
+        // We can use cache if the prototypes match and the Type hasn't been collected
+        if (cachedProto == prototype && cachedType != nullptr)
```

Comment:
If cached type is null, consider just nulling out cache[index] so that those weakrefs can be collected (since that index is basically going to be a cache miss from now on)